### PR TITLE
Add s390x support for CI scripts

### PR DESCRIPTION
**Pull request by jschintag**
Fixes: #1129

Further Todos:
- build and install qemu will be implemented as soon as the configure hypervisor script supports s390x
- currently the Hot Plug CPU Feature is not working on s390x

### CODE REVIEW

File: .ci/s390x/clean_up_s390x.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+#
+# Copyright (c) 2018 IBM
```
Comment (jodh-intel) [inline .ci/s390x/clean_up_s390x.sh]: 2019 (here and below)?

File: .ci/s390x/configuration_s390x.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#
+# Copyright (c) 2018 IBM
+#
+# SPDX-License-Identifier: Apache-2.0#
```
Comment (jodh-intel) [inline .ci/s390x/configuration_s390x.yaml]: Nit: trailing `#`.

File: .ci/s390x/lib_install_qemu_s390x.sh [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+#
+
+set -e
+
+CURRENT_QEMU_VERSION=$(get_version "assets.hypervisor.qemu.version")
+PACKAGED_QEMU="qemu"
+
+get_packaged_qemu_version() {
+        if [ "$ID" == "ubuntu" ]; then
+		#output redirected to /dev/null
```
Comment (jodh-intel) [inline .ci/s390x/lib_install_qemu_s390x.sh]: I'd drop this comment tbh.

File: .ci/s390x/lib_install_qemu_s390x.sh [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+# SPDX-License-Identifier: Apache-2.0
+#
+
+set -e
+
+CURRENT_QEMU_VERSION=$(get_version "assets.hypervisor.qemu.version")
+PACKAGED_QEMU="qemu"
+
+get_packaged_qemu_version() {
+        if [ "$ID" == "ubuntu" ]; then
```
Comment (jodh-intel) [inline .ci/s390x/lib_install_qemu_s390x.sh]: Idea: I'd be tempted to deal with the error scenario first, so in each function:

```sh
[ "$ID" == "ubuntu" ] || die "Unsupported distro: $ID"

 :
```

That will also simplify the remainder of the code as you can remove one layer of `if` testing. Might be worth putting that `die` test at the top global level to avoid having to duplicate either the code in each function, or duplication a call to a function that runs that test?

File: .ci/setup.sh [hunk 5]
```diff
@@ -70,6 +70,16 @@ enable_nested_virtualization() {
 	aarch64 | ppc64le)
 		info "CI running in bare machine"
 		;;
+	s390x)
+		if [ $(systemd-detect-virt) == "kvm" ]; then
+			info "Running in kvm guest. Nested KVM is needed"
+			sudo modprobe -r kvm
+			sudo modprobe kvm nested=1
+			if [ grep -q "N" /sys/module/kvm/parameters/nested 2>/dev/null ]; then
+				die "Failed to find or enable Nested virtualization"
```
Comment (jodh-intel) [inline .ci/setup.sh]: This block looks almost identical to `enable_nested_virtualization()` in `.ci/setup.sh`. Could you just update that to handle os390x?

File: .ci/s390x/configuration_s390x.yaml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+#
+test:
+  - functional
+  - docker
+
+# Same model as is for aarch64
+docker:
```
Comment (jodh-intel) [inline .ci/s390x/configuration_s390x.yaml]: Comparing these files with the `aarch64` ones, they are *very* similar. I can see that there are small differences but I'm wondering if we can create something like:

```
.ci/generic/*
```

... and have `.ci/aarch64/*` and `.ci/s390x/*` include the generic files if possible. That should work for the shell scripts. I guess it won't work as-is for the YAML as that language doesn't support including natively. But, we could either consider breaking the YAML into chunks and reconstructing or having the scripts use tooling to convert the generic YAML files into arch-specific ones?

Bottom line: it's a maintenance burden on the architecture maintainers to ensure multiple very similar files remain updated. But maybe these files won't change much?

Comment (alicefr) [reply .ci/s390x/configuration_s390x.yaml]: hi @jodh-intel now the config looks pretty similar to arm, but on the other hand this file lists (or better black list) what's not working/have been tested on s390x. It could be that in the future the 2 architectures files diverge. On s390x for example, qemu hasn't supported memory hotplug  yet. On s390x, the CPU hotplug has probably a bug with Kata but it should be supported. The file could be a bit redundant but it clearly list what has been tested on s390x. Otherwise, we can have a discussion with @Pennyzct and @nitkon how to integrate the tested features in a more reasonable way for all the arch :)

Comment (jodh-intel) [reply .ci/s390x/configuration_s390x.yaml]: Hi @alicefr - Yes, let's leave these files as separate arch-specific ones for now. We can always re-assess the situation again in the future.

File: .ci/s390x/clean_up_s390x.sh (hunk 1 above)
Comment (jschintag) [reply .ci/s390x/clean_up_s390x.sh]: Thank you, i just copied them and forgot to change the year

File: .ci/setup.sh (hunk 5 above)
Comment (jschintag) [reply .ci/setup.sh]: sure, i can do that

File: .ci/s390x/lib_install_qemu_s390x.sh (hunk 4 above)
Comment (jschintag) [reply .ci/s390x/lib_install_qemu_s390x.sh]: Wouldn't this just mean that i check the content of $ID at two places? I would instead use a switch case to replace this if elif structures and have the error scenario as a default.

Comment (jodh-intel) [reply .ci/s390x/lib_install_qemu_s390x.sh]: Well, as I say, you could make this a global check in `.ci/s390x/lib_install_qemu_s390x.sh`. Something like:

```sh
[ "$ID" == "ubuntu" ] || die "Unsupported distro: $ID"

get_packaged_qemu_version() {
   # ...
}
```

I think that should work since the caller already sources `/etc/os-release` and `lib.sh` which gives access to `die()`.

Comment (jschintag) [reply .ci/s390x/lib_install_qemu_s390x.sh]: Yes, it should work at that position. I changed it in the code.

File: .ci/s390x/filter_docker_s390x.sh [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	local array_docker=$("${GOPATH_LOCAL}/bin/yq" read "${test_config_file}" "${dependency}")
+	[ "${array_docker}" = "null" ] && return
+	mapfile -t _array_docker <<< "${array_docker}"
+	for entry in "${_array_docker[@]}"
+	do
+		_skip_options+=("${entry#- }|")
+	done
+}
+
+main()
```
Comment (grahamwhaley) [inline .ci/s390x/filter_docker_s390x.sh]: Similar to some other items, much of this filter code looks arch generic - at some point we should probably pull a bunch of this out into library functions.
Does not have to block this PR, but will /cc @nitkon @GabyCT on it.

File: .ci/setup.sh [hunk 8]
```diff
@@ -57,11 +57,16 @@ install_docker() {
 
 enable_nested_virtualization() {
 	case "$arch" in
-	x86_64)
+	x86_64 | s390x)
 		if [ "$CI" == true ] && grep -q "N" /sys/module/kvm_intel/parameters/nested 2>/dev/null; then
```
Comment (grahamwhaley) [inline .ci/setup.sh]: given you are doing a non-intel modprobe below, is this `kvm_intel` enabled check also not valid for s390x ?

File: integration/docker/docker.go [hunk 9]
```diff
@@ -34,10 +34,7 @@ const (
 	DebianImage = "debian"
 
 	// FedoraImage is the fedora image
-	FedoraImage = "fedora"
-
-	// CentosImage is the centos image
-	CentosImage = "centos"
+        FedoraImage = "fedora"
```
Comment (grahamwhaley) [inline integration/docker/docker.go]: Is this an intentional change - unclear to me (might have been clearer if it was in its own commit ;-) )

File: integration/docker/docker_arch_base.go [hunk 10]
```diff
@@ -0,0 +1,6 @@
+//+build !s390x
```
Comment (grahamwhaley) [inline integration/docker/docker_arch_base.go]: I don't think we generally use in-file golang build constraints in kata - the constraint via the filename _extention is our normal way. wdty @jodh-intel ?

**Comment by grahamwhaley** [comment]
/test
coz I want to see how it flies....

**Comment by chavafg** [comment]
I see `.ci/.install_qemu.sh.swp`, which shouldn't be part of this change

File: .ci/s390x/lib_install_qemu_s390x.sh [hunk 11]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
+#
+# Copyright (c) 2019 IBM
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+set -e
+
+CURRENT_QEMU_VERSION=$(get_version "assets.hypervisor.qemu.version")
```
Comment (chavafg) [inline .ci/s390x/lib_install_qemu_s390x.sh]: seems like this is not being used, so it can be removed until we use it for building qemu.

File: .ci/install_qemu.sh [hunk 12]
```diff
@@ -91,25 +91,34 @@ if [ -f "${cidir}/${QEMU_ARCH}/lib_install_qemu_${QEMU_ARCH}.sh" ]; then
...
+			packaged_qemu_version=$(get_packaged_qemu_version)
+			short_current_qemu_version=${CURRENT_QEMU_VERSION#*-}
+			if [ "$packaged_qemu_version" == "$short_current_qemu_version" ] && [ -z "${CURRENT_QEMU_COMMIT}" ]; then
+				install_packaged_qemu || build_and_install_qemu
+			else
+				build_and_install_qemu
+			fi
+			;;
+		"s390x")
+			install_packaged_qemu
```
[the remark this answers is not in the document]
Comment (jschintag) [reply .ci/install_qemu.sh]: No, it will not build the source. As i've written in the PR-Message, building Qemu from source is currently not implemented. It will be added later, as soon as the configure_hypervisor.sh script from packaging supports s390x. The code will then look similar to x86_64.

File: .ci/setup.sh (hunk 8 above)
Comment (jschintag) [reply .ci/setup.sh]: Yes it isn't. I overlooked that one.

**Comment by jschintag** [comment]
I added the feedback from @grahamwhaley and @chavafg

File: .ci/install_qemu.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (alicefr) [reply .ci/install_qemu.sh]: @grahamwhaley if this is ok for you we can add the qemu building later. Otherwise, this PR is blocked by the missing support in [configure-hypervisor.sh](https://github.com/kata-containers/packaging/blob/master/scripts/configure-hypervisor.sh). As we obtained the CI integrated, we thought to open the pr to be able at least to install the dependencies and build Kata.

File: .ci/setup.sh [hunk 13]
```diff
@@ -57,12 +57,15 @@ install_docker() {
 
 enable_nested_virtualization() {
 	case "$arch" in
-	x86_64)
-		if [ "$CI" == true ] && grep -q "N" /sys/module/kvm_intel/parameters/nested 2>/dev/null; then
+	x86_64 | s390x)
+		kvm_arch="kvm"
+		[ ${arch} == "x86_64" ] && kvm_arch="kvm_intel"
+		echo "${kvm_arch}"
```
Comment (alicefr) [inline .ci/setup.sh]: Do we need this print?

Comment (jschintag) [reply .ci/setup.sh]: no we don't

File: .ci/install_qemu.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (grahamwhaley) [reply .ci/install_qemu.sh]: Sure, its fine - was just checking...

**Comment by grahamwhaley** [comment]
Hi @jschintag - thanks for the updates. We are getting there :-)
The CI check is moaning that you have no text body to your commit though - can you add a brief description (or, as full as you like ;-) ) pls.

```
Found 1 commit between commit 5ea2931de74c090c3246ab52db64bcc59e32177f and branch master
ERROR: Commit 5ea2931de74c090c3246ab52db64bcc59e32177f: single-line "Signed-off-by" body not permitted
ERROR: checkcommits failed. See the document below for help on formatting
commits for the project.
https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format
```

File: integration/docker/docker_arch_base.go (hunk 10 above)
Comment (jodh-intel) [reply integration/docker/docker_arch_base.go]: Yes, but you can't negate using the filename afaik, so this one is fine.

File: integration/docker/docker_s390x.go [hunk 14]
```diff
@@ -0,0 +1,8 @@
+//+build s390x
```
Comment (jodh-intel) [inline integration/docker/docker_s390x.go]: Nit: not required as you've already encoded the platform in the name of the file itself.

**Comment by jschintag** [comment]
@grahamwhaley  Ok, i added a body to the commit message. Hopefully the CI check likes it now.
@jodh-intel I also removed the build constraint header in integration/docker/docker_s390x.go

**Comment by grahamwhaley** [comment]
Hi @jschintag - CI happy with the commit now, but it has noticed two files don't have the SPDX copyright headers...

```
All commit checks passed.
INFO: Checking for SPDX license headers
ERROR: Required license identifier ('SPDX-License-Identifier: Apache-2.0') missing from following files:
integration/docker/docker_arch_base.go
integration/docker/docker_s390x.go
The command "bash .ci/static-checks.sh github.com/kata-containers/tests" exited with 1.
```

I know, battling the CIs... but, I'm hoping we are nearly there!

**Comment by jschintag** [comment]
@grahamwhaley No problem, i'll fix all the Problems the CI has.

**Comment by grahamwhaley** [comment]
side note - @ttx, look, Zuul SoB/WIP checks active and working on this tests repo PR :-)

**Comment by grahamwhaley** [comment]
@jschintag thanks - Travis now happy, along with the new Zuul checks (that replace some of the broken pullapprove CI). I'll fire the tests again so we can check we have not broken any other arch's...

**Comment by grahamwhaley** [comment]
From the 16.04 CI, looks like we have some code issue:

```
14:39:05 bash sanity/check_sanity.sh
14:39:05 ./ginkgo -v -focus "" -skip "" ./integration/docker/ -- -runtime=kata-runtime -timeout=60
14:39:05 Failed to compile docker:
14:39:05 
14:39:05 can't load package: package github.com/kata-containers/tests/integration/docker: 
14:39:05 integration/docker/docker_arch_base.go:8:1: expected 'package', found 'const'
14:39:05 integration/docker/docker_arch_base.go:10:9: expected ';', found CentosImage
```

**Comment by jschintag** [comment]
I fixed the issue

**Comment by jodh-intel** [comment]
Thanks @jschintag.

/retest

**Comment by jschintag** [comment]
@grahamwhaley @jodh-intel Since https://github.com/kata-containers/packaging/pull/333 was merged i added the support for building qemu on s390x.

File: integration/docker/docker_arch_base.go [hunk 15]
```diff
@@ -0,0 +1,12 @@
+// +build !s390x
+
+// Copyright (c) 2019 IBM
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+package docker
```
Comment (grahamwhaley) [inline integration/docker/docker_arch_base.go]: odd, I think the CI is moaning about this, but did not moan about the one in docker_s390x.go?
```
integration/docker/docker_arch_base.go:3:1:warning: package comment should be of the form "Package docker ..." (golint)
The command "bash .ci/static-checks.sh github.com/kata-containers/tests" exited with 1.
```

@sameo - is this maybe at all related to the static check stuff you were looking at yesterday? (that is, not that you broke it, but maybe you know if there is something related to what you were debugging/fixing etc.).

Comment (jschintag) [reply integration/docker/docker_arch_base.go]: It seems the CI wants at least one empty line between the package and the License

**Comment by jodh-intel** [comment]
@chavafg - the firecracker CI is failing but that doesn't appear to be related to this PR. Can we just merge this as all other CI's are passing?

**Comment by chavafg** [comment]
sure, lets merge, but I think this needs 1 more ack

**Comment by grahamwhaley** [comment]
Only fc failing - let's merge... and, let's figure out what the deal with fc is...
